File: clients/go/workshop/session1-basic/starter/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
	"time"
)

func main() {
	// TODO: Create a new basic client
	// Hint: Use NewClient("", 120*time.Second) to use default URL from constant
	client := NewClient("", 120*time.Second)
	defer client.Close()

	// TODO: Check server health
	fmt.Println("Checking server health...")
	// Hint: Call client.CheckHealth() and handle the error
	health, err := client.CheckHealth()
	if err != nil {
		log.Fatalf("Health check failed: %v", err)
	}

	// TODO: Print the health status
	fmt.Printf("Server Status: %s\n", health.Status)
	fmt.Printf("Ollama Accessible: %v\n", health.OllamaAccessible)
	if health.Message != nil {
		fmt.Printf("Message: %s\n", *health.Message)
	}
	fmt.Println()

	// Interactive question loop
	fmt.Println("=== AISHE Question Answering (Session 1: Basic Client) ===")
	fmt.Println("Type your question and press Enter. Type 'exit' to quit.")
	fmt.Println()

	scanner := bufio.NewScanner(os.Stdin)

	for {
		fmt.Print("Question: ")
		if !scanner.Scan() {
			break
		}

		question := strings.TrimSpace(scanner.Text())

		// Check for exit command
		if strings.ToLower(question) == "exit" {
			fmt.Println("Goodbye!")
			break
		}

		// Skip empty questions
		if question == "" {
			continue
		}

		// TODO: Measure execution time and ask the question
		// Hint: Use time.Now() before calling AskQuestion, then time.Since(start).Seconds()
		start := time.Now()
		// TODO: Call client.AskQuestion(question) here
		// answer, err := client.AskQuestion(question)
		executionTime := time.Since(start).Seconds()

		// TODO: Handle errors
		// if err != nil {
		//     fmt.Printf("❌ Error: %v\n\n", err)
		//     continue
		// }

		// TODO: Display results
		// Expected output format:
		//
		// Answer: <the answer>
		//
		// Source: AISHE API
		// Execution Time: X.XX seconds
		//
		// Wikipedia Sources:
		//   [1] <title>
		//       <url>
		//

		fmt.Println()
		fmt.Println("TODO: Display the answer here")
		fmt.Printf("Execution Time: %.2f seconds\n", executionTime)
		fmt.Println()
	}

	if err := scanner.Err(); err != nil {
		log.Fatalf("Error reading input: %v", err)
	}
}

File: clients/js/workshop/session1-basic/starter/main.ts
import readline from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";

import type { AnswerResponse, HealthResponse } from "aishe-client";

import { AIsheHTTPClient } from "./client.js";

async function main(): Promise<void> {
    // TODO: create a new AIsheHTTPClient instance
    // ..........................

    // TODO: check AIshe's health
    // Hint: you'll need to use the 'await' operator with async functions.
    console.log("Checking AIshe's health...");
    // ..........................

    // TODO: print the health status
    // Hint: you need to print `status`, `ollama_accessible`, and `message` if it exists.
    // ..........................

    // Interactive question loop
    console.log("=== AIshe Question Answering (Session 1: Basic Client) ===");
    console.log("Type your question and press Enter to get an answer.");
    console.log("Enter CTRL+C, 'quit', 'exit', or 'q' to exit.");
    console.log();

    const rl = readline.createInterface({ input, output });
    // Handle CTRL+C at readline level
    rl.on("SIGINT", () => {
        console.log("\nReceived CTRL+C. Exiting...");
        rl.close();
    });

    while (true) {
        const question = (await rl.question("Your question: ")).trim();
        if (["quit", "exit", "q"].includes(question.toLowerCase())) {
            console.log("\nGoodbye!");
            break;
        }
        // Skip empty questions
        if (!question) {
            continue;
        }

        // TODO: ask AIshe a question, handle errors, measure execution time
        // Hint: use performance for measuring execution time
        // Hint: performance measures in milliseconds, so you need to convert it to seconds
        // ..........................

        // TODO: dispaly results
        // Expected output format:
        //
        // Answer: <answer>
        // Source: AIshe API
        //
        // Processing time: <processing_time>
        // Measured execution time: <measured_time>
        //
        // Wikipedia sources:
        //   [1] <title>
        //       <url>
        //   [2] <title>
        //       <url>
        //   [3] <title>
        //       <url>

        // ..........................
    }
    rl.close();
}

// Run only when executed directly
if (process.argv[1] && import.meta.url.endsWith(process.argv[1])) {
    main().catch((err) => {
        console.error("Fatal error:", err);
        process.exit(1);
    });
}
